omx/gstomxaacdec.c

```diff
@@ -96,70 +94,114 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
   const gchar *stream_format;
 
   gst_omx_port_get_port_definition (port, &port_def);
-  port_def.format.audio.eEncoding = OMX_AUDIO_CodingAAC;
+
+  if (port->index == GST_OMX_AAC_DEC_INPUT_PORT) {
+    port_def.nPortIndex = port->index;
+    port_def.eDir = OMX_DirInput;
+    port_def.nBufferCountActual = 1;
+    port_def.nBufferCountMin = 1;
+    port_def.nBufferSize = GST_OMX_AAC_DEC_INPUT_PORT_BUFFERSIZE;
+    port_def.bEnabled = OMX_TRUE;
+    port_def.bPopulated = OMX_FALSE;
+    port_def.eDomain = OMX_PortDomainAudio;
+    port_def.bBuffersContiguous = OMX_FALSE;
+    port_def.nBufferAlignment = 32;
+    port_def.format.audio.cMIMEType = (OMX_STRING)"ADEC";
+    port_def.format.audio.pNativeRender = NULL;
+    port_def.format.audio.eEncoding = OMX_AUDIO_CodingAAC;
+    port_def.format.audio.bFlagErrorConcealment = OMX_FALSE;
+    GST_DEBUG_OBJECT (self, "Updating input port definition");
+
+  } else if (port->index == GST_OMX_AAC_DEC_OUTPUT_PORT) {
+    port_def.nPortIndex = port->index;
+    port_def.eDir = OMX_DirOutput;
+    port_def.nBufferCountActual = 1;
+    port_def.nBufferCountMin = 1;
+    port_def.nBufferSize = GST_OMX_AAC_DEC_OUTPUT_PORT_BUFFERSIZE;
+    port_def.bEnabled = OMX_TRUE;
+    port_def.bPopulated = OMX_FALSE;
+    port_def.eDomain = OMX_PortDomainAudio;
+    port_def.bBuffersContiguous = OMX_FALSE;
+    port_def.nBufferAlignment = 32;
+    port_def.format.audio.cMIMEType = (OMX_STRING)"PCM";
+    port_def.format.audio.pNativeRender = NULL;
+    port_def.format.audio.bFlagErrorConcealment = OMX_FALSE;
+    port_def.format.audio.eEncoding = OMX_AUDIO_CodingUnused;
+    GST_DEBUG_OBJECT (self, "Updating output port definition");
+  }
+
   err = gst_omx_port_update_port_definition (port, &port_def);
   if (err != OMX_ErrorNone) {
     GST_ERROR_OBJECT (self,
-        "Failed to set AAC format on component: %s (0x%08x)",
-        gst_omx_error_to_string (err), err);
+        "Failed to update port %d definition with AAC format: %s (0x%08x)",
+        port->index, gst_omx_error_to_string (err), err);
     return FALSE;
   }
 
-  GST_OMX_INIT_STRUCT (&aac_param);
-  aac_param.nPortIndex = port->index;
+  /* Set AAC params using sink pad caps */
+  if (port->index == GST_OMX_AAC_DEC_INPUT_PORT && caps != NULL ) {
```

Comment:
If `caps == NULL` shouldn't you return an error?  If there are no caps, then the encoder parameters will not be able to be set.

Reply:
@michaelgruner That applies only for the omx input port configuration, for the omx output port configuration they are not necessary.

Reply:
I don't understand. So, is it possible for an input port to don't have caps? is that a normal case and not an error?

Reply:
@michaelgruner Sorry my answer was not clear.  >> Q: If caps == NULL shouldn't you return an error? >> A: Yes you are correct. I removed that condition and placed the proper pointer guards for caps/pointer checking in the class methods. 80f12f3b53c13492fb3b54d22b206df13695b772